cmd/hub/mcscontroller/controller.go

```diff
@@ -117,16 +124,138 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 // - ServiceImports are owned by the corresponding ServiceExport and are thus
 //	 automatically deleted by k8s when their ServiceExport is deleted.
 //
-func (r *ServiceExportReconciler) reconcile(ctx context.Context, name types.NamespacedName) (ctrl.Result, error) {
-	_ = r.Log.WithValues("service", name)
+func (r *ServiceExportReconciler) reconcile(ctx context.Context, name types.NamespacedName, listOptions *client.ListOptions) (ctrl.Result, error) {
+	log := r.Log.WithValues("service", name)
+	var exports mcsv1a1.ServiceExportList
+	if err := r.Client.List(ctx, &exports, listOptions); err != nil {
+		log.Error(err, "unable to list service's service export")
+		return ctrl.Result{}, err
+	}
+	primary, err := getPrimaryExportObject(exports)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	for _, se := range exports.Items {
+		if se.DeletionTimestamp != nil {
+			continue
+		}
 
-	//------------------------------------------
-	// -- @TODO CODE ABOVE SPECIFIED BEHAVIOR --
-	//------------------------------------------
+		exportSpec := &mcs.ExportSpec{}
+		err := exportSpec.UnmarshalObjectMeta(&se.ObjectMeta)
+		if err != nil {
+			log.Info("Failed to unmarshal exportSpec from serviceExport")
+			return ctrl.Result{}, err
+		}
 
+		err = exportSpec.IsCompatibleWith(primary)
+		if err != nil {
+			lhutil.UpdateServiceExportConditions(ctx, &se, log, mcsv1a1.ServiceExportConflict, corev1.ConditionTrue, "", err.Error())
+			r.Client.Status().Update(ctx, &se)
+			si := mcsv1a1.ServiceImport{}
+			err = r.Client.Get(ctx, name, &si)
+			if err == nil {
+				r.Client.Delete(ctx, &si)
+			} else {
+				if !apierrors.IsNotFound(err) {
+					log.Error(err, "Failed to get the service Import to delete")
+					return ctrl.Result{}, err
+				}
+			}
+		} else {
+			lhutil.UpdateServiceExportConditions(ctx, &se, log, mcsv1a1.ServiceExportConflict, corev1.ConditionFalse, "", "successfully update the service export")
+			err = r.Client.Status().Update(ctx, &se)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			r.ensureImportFor(ctx, &se)
+		}
+	}
 	return ctrl.Result{}, nil
 }
 
+// getPrimaryExportObject finds the Primary ExportSpec from ServiceExportList.
+// Retruns the ExportSpec object that determined as the primary one from
+// the given ServiceExportList according to the conflict resolution specification set in the
+// KEP, that is implemented in IsPreferredOver method of ExportSpec.
+func getPrimaryExportObject(exportList mcsv1a1.ServiceExportList) (*mcs.ExportSpec, error) {
+	var primary *mcs.ExportSpec
+	for i, se := range exportList.Items {
```

Comment:
should return error in case of an empty list?

Reply:
I wasn't sure about it,
currently, it returns a "nil" if the list is empty, which can make sense - if the list is empty, then the "primary object" there is nil (there isn't one).
If you think it would be better to raise an error I can add it.

Reply:
nil is ok too, jsut make sure that you handle this case - you should then produce an error in the caller..

Reply:
I think explicit error is better in this case